onnxruntime/contrib_ops/cpu/quantization/matmul_nbits.cc

```diff
@@ -322,7 +322,8 @@ Status MatMulNBits::Compute(OpKernelContext* ctx) const {
   if (has_single_b_matrix) {
     const auto compute_type = static_cast<MLAS_SQNBIT_GEMM_COMPUTE_TYPE>(accuracy_level_);
 
-    if (MlasIsSQNBitGemmAvailable(nbits_, block_size_, compute_type)) {
+    // mlas nbits implementation requires packed b. update this logic if it changes.
```

Comment:
the way to check whether it requires packed B is `MlasSQNBitGemmPackQuantBDataSize() > 0`.

if you want to require that B is packed here, you can simplify the logic below that either reads from `packed_b_` or `ctx->Input<Tensor>(1)`. it currently handles both packed/not packed cases.